File: dijkstra/Heap.java
package dijkstra;

public class Heap 
{
    private int[]   heap;
    private int[]   values;
    private int[]   indices;
    private boolean maxheap;
    private int     N;
    private int     max_elems;
	
    Heap(int max_elems, boolean maxheap)
    {
        this.N         = 0;
        this.max_elems = max_elems;
        this.heap      = new int[max_elems + 1];
        this.values    = new int[max_elems + 1];
        this.indices   = new int[max_elems + 1];
        this.maxheap   = maxheap;
		
        for (int i = 0; i < max_elems; i++)
	{
            this.heap[i]    = -1;
            this.values[i]  = -1;
            this.indices[i] = -1;
	}
    }

    public boolean contains(int item)
    {
	return this.indices[item] != -1;
    }

    public boolean isEmpty() 
    {
        return this.N == 0;
    }
	
    private void swap(int i, int j)
    {
        int tmp;
		
	tmp = this.heap[i];
	this.heap[i] = this.heap[j];
	this.heap[j] = tmp;
	
	this.indices[this.heap[i]] = i;
	this.indices[this.heap[j]] = j;
    }

    private boolean isPrior(int i, int j)
    {
        if (this.maxheap)
            return this.values[this.heap[i]] > this.values[this.heap[j]];
	else
            return this.values[this.heap[i]] < this.values[this.heap[j]];
    }

    public void bubbleUp(int pos)
    {
        int parent;
		
	parent = pos / 2;
        while (parent >= 1 && isPrior(pos, parent))
	{
            swap(pos, parent);

            pos = parent;
            parent = pos / 2;
        }
    }
            
    public void bubbleDown(int pos)
    {
        int left;
	int right;
        int m;
        
        left = pos * 2;
        while (left < this.N)
	{
            right = left + 1;
            if (right < this.N && isPrior(right, left))
                m = right;
            else
                m = left;

            if (!isPrior(m, pos))
                break;
            
            swap(pos, m);
            
            pos = m;
            left = pos * 2;
	}
    }	
	
    public void insert(int item, int val)
    {
        if (this.N + 1 <= max_elems)
	{
            this.N++;
            this.heap[this.N] = item;
            this.indices[item] = this.N;
            this.values[item] = val;
            bubbleUp(this.N);
        }
    }
	
    public int remove()
    {
        int m;
		
        if (isEmpty())
            return -1;
        
        m = this.heap[1];
		
        swap(1, this.N);
        this.N--;
		
        bubbleDown(1);
	
        this.indices[m]  = -1;
        this.heap[N + 1] = -1;

        return m;
    }

    public void update(int item, int val)
    {
        int pos;
		
        this.values[item] = val;
		
        pos = this.indices[item];
        bubbleUp(pos);
    }
}

File: dijkstra/heap.py
# an indexed priority queue

from operator import gt, lt

class Heap:
    heap = [None]
    value = {}
    indices = {}            # keep track of an item's index in the list representing the tree
    comparator = None

    def __init__(self, maxq=False):
        if maxq:
            self.comparator = gt
        else:
            self.comparator = lt

    def contains(self, item):
        return item in self.indices     # constant-time operation
        
    def empty(self):
        return len(self.heap) == 1      # the list only has the dummy item in it

    # given two item's indices, swap their positions
    def swap(self, i, j):
        self.heap[i], self.heap[j] = self.heap[j], self.heap[i]
        self.indices[self.heap[i]] = i
        self.indices[self.heap[j]] = j

    # given two item's indices, 
    # decide whether the item in the i-th position of the tree/list is prior
    def is_prior(self, i, j):
        return self.comparator(self.value[self.heap[i]], self.value[self.heap[j]])

    # given an item's index, try to percolate the item up the tree
    def bubble_up(self, pos):
        parent = pos // 2

        while parent >= 1 and self.is_prior(pos, parent):
            self.swap(pos, parent)

            pos = parent
            parent = pos // 2
            
    # given an item's index, try to percolate down the tree
    def bubble_down(self, pos):
        left = pos * 2

        while left < len(self.heap):
            # of the node's two children, take the one with the higher priority
            right = left + 1
            if right < len(self.heap) and self.is_prior(right, left):
                m = right
            else:
                m = left

            # if item isn't less prior, it can stay where it is
            if not self.is_prior(m, pos):
                break
            
            self.swap(pos, m)
            
            pos = m
            left = pos * 2

    # place an item of some value in the tree
    def insert(self, item, val):
        # place the item in the tree and set its value
        self.heap.append(item)
        self.value[item] = val
        
        # set the item's index
        pos = len(self.heap) - 1
        self.indices[item] = pos

        # sort out the new order
        self.bubble_up(pos)

    # remove and return the top-priority element
    def remove(self):
        if self.empty():
            return None

        # remove the top-priority element
        m = self.heap[1]
        self.value.pop(m)
        self.indices.pop(m)
        
        # replace top-priority item with lowest priority item
        # sort out the new order
        u = self.heap.pop()         # lowest priority item is at the end of the list
        if not self.empty():        # top-priority item could be lowest priority item
            self.heap[1] = u
            self.indices[u] = 1
            self.bubble_down(1)

        return m

    # update an item's value
    def update(self, item, val):
        self.value[item] = val
        pos = self.indices[item]
        self.bubble_up(pos)         # bubble_up() assumes the item's priority increases
